File: src/android/KahunaPlugin.java
/*
 * Kahuna CONFIDENTIAL
 * __________________
 *
 *  2014 Kahuna, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Kahuna, Inc. and its suppliers,
 * if any.  The intellectual and technical concepts contained
 * herein are proprietary to Kahuna, Inc.
 * and its suppliers and may be covered by U.S. and Foreign Patents,
 * patents in process, and are protected by trade secret or copyright law.
 * Dissemination of this information or reproduction of this material
 * is strictly forbidden unless prior written permission is obtained
 * from Kahuna, Inc.
 */
package com.kahuna.phonegap.sdk;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;

import com.kahuna.sdk.KahunaAnalytics;
import com.kahuna.sdk.KahunaUserCredentialKeys;

public class KahunaPlugin extends CordovaPlugin {
	
	private static final String TAG = "Kahuna";
	
	private static final String START_WITH_KEY = "startWithKey";
	private static final String ON_START = "onStart";
	private static final String ON_STOP = "onStop";
	private static final String TRACK_EVENT = "trackEvent";
	private static final String SET_USER_CREDENTIALS = "setUserCredentials";
	private static final String LOGOUT = "logout";
	private static final String SET_USER_ATTRIBUTES = "setUserAttributes";
	private static final String ENABLE_PUSH = "enablePush";
	private static final String DISABLE_PUSH = "disablePush";
	private static final String SET_DEBUG_MODE = "setDebugMode";
	
	@Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
		
		try {
			if(START_WITH_KEY.equals(action)) {
				String secretKey = args.getString(0);
				String senderId = null;
				
				if(args.length() > 1) {
					senderId = args.getString(1);
				}
				
				this.startWithKey(secretKey, senderId);
				return true;
			}
			else if(ON_START.equals(action)) {
				this.onStart();
			}
			else if(ON_STOP.equals(action)) {
				this.onStop();
			}
			else if(TRACK_EVENT.equals(action)) {
				String eventName = args.getString(0);
				this.trackEvent(eventName);
			}
			else if(SET_USER_ATTRIBUTES.equals(action)) {
				Map<String, String> userAttributes = null;
				JSONObject userAttsObject = args.getJSONObject(0);
				userAttributes = new HashMap<String, String>();
				Iterator<?> keys = userAttsObject.keys();
				while (keys.hasNext()) {
					String key = (String) keys.next();
					userAttributes.put(key, userAttsObject.getString(key));
				}
				this.setUserAttributes(userAttributes);
			}
			else if(SET_USER_CREDENTIALS.equals(action)) {
				String username = args.getString(0);
				String email = args.getString(1);
				this.setUserCredentials(username, email);
			}
			else if(LOGOUT.equals(action)) {
				this.logout();
			}
			else if(ENABLE_PUSH.equals(action)) {
				this.enablePush();
			}
			else if(DISABLE_PUSH.equals(action)) {
				this.disablePush();
			}
			else if(SET_DEBUG_MODE.equals(action)) {
				boolean enabled = args.getBoolean(0);
				this.setDebugMode(enabled);
			}
			else {
				Log.w(TAG, "Execute " + action + " did no match any method definition defined by Kahuna.");
				return false;
			}
		}
		catch(Exception e) {
			e.printStackTrace();
			return false;
		}
        return false;
    }
    
    private void startWithKey(final String secretKey, final String senderId) {
    	KahunaAnalytics.onAppCreate(this.cordova.getActivity().getApplicationContext(), secretKey, senderId);
    	KahunaAnalytics.setHybridSDKVersion("cordova_sdk_version", "437");
    	KahunaAnalytics.start();
    }
    
    private void onStart() {
    	KahunaAnalytics.start();
    }
    
    private void onStop() {
    	KahunaAnalytics.stop();
    }
    
    private void trackEvent(final String eventName) {
    	KahunaAnalytics.trackEvent(eventName);
    }
    
    private void setUserCredentials(final String username, final String email) {
    	if(username != null)
    		KahunaAnalytics.setUserCredential(KahunaUserCredentialKeys.USERNAME_KEY, username);
    	if(email != null)
    		KahunaAnalytics.setUserCredential(KahunaUserCredentialKeys.EMAIL_KEY, email);
    }
    
    private void logout() {
    	KahunaAnalytics.logout();
    }
    
    private void setUserAttributes(Map<String, String> userAttributes) {
    	KahunaAnalytics.setUserAttributes(userAttributes);
    }
    
    private void enablePush() {
    	KahunaAnalytics.enablePush();
    }
    
    private void disablePush() {
    	KahunaAnalytics.disablePush();
    }
    
    private void setDebugMode(boolean enable) {
    	KahunaAnalytics.setDebugMode(enable);
    }
    
    public class KAPushReceiver extends BroadcastReceiver {

    	@Override
    	public void onReceive(Context context, Intent intent) {
    		String action = intent.getAction();
    		if(action.equals(KahunaAnalytics.ACTION_PUSH_CLICKED)){
    			/* Push notification was clicked. */		
    			Bundle extras = intent.getBundleExtra(KahunaAnalytics.EXTRA_LANDING_DICTIONARY_ID);
    			for (String key : extras.keySet()) {
    			    Object value = extras.get(key);
    			    Log.d(TAG, String.format("%s : %s (%s)", key,  
    			        value.toString(), value.getClass().getName()));
    			}
    		}
    		if(action.equals(KahunaAnalytics.ACTION_PUSH_RECEIVED)){
    			/* Push notification was received. */
    			Log.i(TAG, "Received Kahuna push");
    			
    			Bundle extras = intent.getBundleExtra(KahunaAnalytics.EXTRA_LANDING_DICTIONARY_ID);
    			for (String key : extras.keySet()) {
    			    Object value = extras.get(key);
    			    Log.d(TAG, String.format("%s : %s (%s)", key,  
    			        value.toString(), value.getClass().getName()));
    			}
    		}
    	}

    }
}

File: src/ios/KahunaPlugin.m
/*
 * Kahuna CONFIDENTIAL
 * __________________
 *
 *  2014 Kahuna, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Kahuna, Inc. and its suppliers,
 * if any.  The intellectual and technical concepts contained
 * herein are proprietary to Kahuna, Inc.
 * and its suppliers and may be covered by U.S. and Foreign Patents,
 * patents in process, and are protected by trade secret or copyright law.
 * Dissemination of this information or reproduction of this material
 * is strictly forbidden unless prior written permission is obtained
 * from Kahuna, Inc.
 */
//
//  KahunaCordovaPlugin.m
//

#import "KahunaPlugin.h"
#import "KahunaAnalytics.h"

@implementation KahunaPlugin

static NSString *secretKey;

- (void)startWithKey:(CDVInvokedUrlCommand*)command {
    CDVPluginResult* pluginResult = nil;
    NSString* key = [command.arguments objectAtIndex:0];
    
    if (key != nil && [key length] > 0) {
        [KahunaAnalytics setDeepIntegrationMode:true];
        [KahunaAnalytics setSDKWrapper:@"cordova" withVersion:@""];
        [KahunaAnalytics startWithKey:key];
        secretKey = key;
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    } else {
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Need non-nil & non-empty Secret Key."];
    }
    
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)onStart:(CDVInvokedUrlCommand*)command {
    CDVPluginResult* pluginResult = nil;
    if (secretKey != nil && [secretKey length] > 0) {
        [KahunaAnalytics startWithKey:secretKey];
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    } else {
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Need non-nil & non-empty Secret Key to be set in onStart."];
    }
    
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)onStop:(CDVInvokedUrlCommand*)command {
    // Not implemented in iOS since we have the observer already setup.
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)trackEvent:(CDVInvokedUrlCommand*)command {
    CDVPluginResult* pluginResult = nil;
    NSString* eventName = [command.arguments objectAtIndex:0];
    
    if (eventName != nil && [eventName length] > 0) {
        [KahunaAnalytics trackEvent:eventName];
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    } else {
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Need non-nil & non-empty event name."];
    }
    
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)setUserCredentials:(CDVInvokedUrlCommand*)command {
    CDVPluginResult* pluginResult = nil;
    NSString* username = [command.arguments objectAtIndex:0];
    NSString* email = [command.arguments objectAtIndex:1];
    
    [KahunaAnalytics setUserCredentialsWithKey:KAHUNA_CREDENTIAL_USERNAME andValue:username];
    [KahunaAnalytics setUserCredentialsWithKey:KAHUNA_CREDENTIAL_EMAIL andValue:email];
    
    pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)logout:(CDVInvokedUrlCommand*)command {
    CDVPluginResult* pluginResult = nil;
    
    [KahunaAnalytics logout];
    
    pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)setUserAttributes:(CDVInvokedUrlCommand*)command {
    CDVPluginResult* pluginResult = nil;
    NSDictionary *userAttributes = [command.arguments objectAtIndex:0];
    
    [KahunaAnalytics setUserAttributes:userAttributes];
    
    pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)enablePush:(CDVInvokedUrlCommand*)command {
    CDVPluginResult* pluginResult = nil;
    NSDictionary *notificationOptions = [command.arguments objectAtIndex:0];
    
    NSUInteger notificationTypes = 0;
    id badgeArg = nil;
    id soundArg = nil;
    id alertArg = nil;
    
    if (notificationOptions && [notificationOptions isKindOfClass:[NSDictionary class]])
    {
        badgeArg = [notificationOptions objectForKey:@"badge"];
        soundArg = [notificationOptions objectForKey:@"sound"];
        alertArg = [notificationOptions objectForKey:@"alert"];
    }
    else
    {
        alertArg = @"true";
    }
    
    if ([badgeArg isKindOfClass:[NSString class]])
    {
        if ([badgeArg isEqualToString:@"true"])
            notificationTypes |= (1 << 0);
    }
    else if ([badgeArg boolValue])
        notificationTypes |= (1 << 0);
    
    if ([soundArg isKindOfClass:[NSString class]])
    {
        if ([soundArg isEqualToString:@"true"])
            notificationTypes |= (1 << 1);
    }
    else if ([soundArg boolValue])
        notificationTypes |= (1 << 1);
    
    if ([alertArg isKindOfClass:[NSString class]])
    {
        if ([alertArg isEqualToString:@"true"])
            notificationTypes |= (1 << 2);
    }
    else if ([alertArg boolValue])
        notificationTypes |= (1 << 2);
    
    if (notificationTypes == 0)
    {
        return;
    }

#ifdef NSFoundationVersionNumber_iOS_7_1
    // iOS 8 has a new way to register for notifications. This conditional code takes care of iOS 8 and
    // previous versions.
    if ([[UIApplication sharedApplication] respondsToSelector:@selector(registerForRemoteNotifications)]) {
        UIUserNotificationSettings *settings = [UIUserNotificationSettings settingsForTypes:notificationTypes categories:nil];
        [[UIApplication sharedApplication] registerUserNotificationSettings:settings];
        [[UIApplication sharedApplication] registerForRemoteNotifications];
    } else {
        [[UIApplication sharedApplication] registerForRemoteNotificationTypes:notificationTypes];
    }
#else
    [[UIApplication sharedApplication] registerForRemoteNotificationTypes:notificationTypes];
#endif
    
    
    pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)disablePush:(CDVInvokedUrlCommand*)command {
    // Not implemented in iOS since the OS controls enabling/disabling push
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)setDebugMode:(CDVInvokedUrlCommand*)command {
    CDVPluginResult* pluginResult = nil;
    id value = [command.arguments objectAtIndex:0];
    if (!([value isKindOfClass:[NSNumber class]])) {
        value = [NSNumber numberWithBool:NO];
    }
    
    [KahunaAnalytics setDebugMode:[value boolValue]];
    
    pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

@end
